airflow/models.py

```diff
@@ -3455,14 +3366,176 @@ def find(dag_id, run_id=None, state=None, external_trigger=None, session=None):
+        )
+        if state:
+            if isinstance(state, six.string_types):
+                tis = tis.filter(TI.state == state)
+            else:
+                # this is required to deal with NULL values
+                if None in state:
+                    tis = tis.filter(or_(
+                        TI.state.in_(state),
+                        TI.state.is_(None))
+                    )
+                else:
+                    tis = tis.filter(TI.state.in_(state))
+
+        return tis.all()
+
+    @provide_session
+    def get_task_instance(self, task_id, session=None):
+        """
+        Returns the task instance specified by task_id for this dag run
+        :param task_id: the task id
+        """
+
+        TI = TaskInstance
+        ti = session.query(TI).filter(
+            TI.dag_id == self.dag_id,
+            TI.execution_date == self.execution_date,
+            TI.task_id == task_id
+        ).one()
+
+        return ti
+
+    def get_dag(self):
+        """
+        Returns the Dag associated with this DagRun
+        :param session: database session
+        :return: DAG
+        """
+        if not self.dag:
+            raise AirflowException("{}.dag needs to be set".format(self))
+
+        return self.dag
+
+    @provide_session
+    def update_state(self, session=None):
+        """
+        Determines the overall state of the DagRun based on the state
+        of its TaskInstances.
+        :returns State:
+        """
+
+        dag = self.get_dag()
+        tis = self.get_task_instances(session=session)
+
+        logging.info("Updating state for {} considering {} task(s)".format(self, len(tis)))
+
+        for ti in tis:
+            ti.task = dag.get_task(ti.task_id)
+
+        # pre-calculate
+        # db is faster
+        # todo: this can actually get pretty slow: one task costs between 0.01-015s
+        start_dttm = datetime.now()
+        unfinished_tasks = self.get_task_instances(
+            state=State.unfinished(),
+            session=session
+        )
+        has_unfinished_tasks = True if unfinished_tasks else False
+        none_depends_on_past = all(t.task.depends_on_past for t in unfinished_tasks)
+        no_dependencies_met = all(not t.are_dependencies_met(session=session)
+                                  for t in unfinished_tasks)
+        duration = (datetime.now() - start_dttm).total_seconds() * 1000
+        Stats.timing("dagrun.dependency-check.{}.{}".
+                     format(self.dag_id, self.execution_date), duration)
+
+        # Runs are considered deadlocked if there are unfinished tasks but
+        # none of them can run. First we check across *all* dagruns in case
+        # there are depends_on_past relationships which could make individual
+        # dags look deadlocked incorrectly. Later we will check individual
+        # dagruns, as long as they don't have depends_on_past=True
+
+        all_deadlocked = (has_unfinished_tasks and no_dependencies_met)
```

Comment:
I'm not sure this logic still works here, since now the only TI's being considered are from a single DagRun.

Let's say there are two DagRuns with a `depends_on_past=True` for the first task. The first DagRun will run normally. I think the second DagRun will report that it has unfinished tasks but that they can't run due to `depends_on_past` and it will deadlock. 

I think we should stop checking for that type of Deadlock, for simplicity. The deadlock in line 3489 will check if just this DAG is deadlocked, though it will do it in a way that respects the weird behavior of depends_on_past.